packages/cache/src/index.ts

```diff
@@ -0,0 +1,170 @@
+import { Redis } from 'ioredis';
+import { LRUCache } from 'lru-cache';
+import { logger } from '@prairielearn/logger';
+import * as Sentry from '@prairielearn/sentry';
+
+class Cache {
+  cacheEnabled = false;
+  cacheType = 'none';
+  memoryCache?: LRUCache<string, string>;
+  client?: Redis;
+  cacheKeyPrefix = 'prairielearn-cache:';
+
+  async init(config: {
+    cacheType: 'none' | 'memory' | 'redis';
+    cacheKeyPrefix: string;
+    redisUrl?: string | null;
+  }) {
+    this.cacheType = config.cacheType;
+    this.cacheKeyPrefix = config.cacheKeyPrefix;
+    if (!this.cacheType || this.cacheType === 'none') {
+      // No caching
+      this.cacheEnabled = false;
+      return;
+    }
+
+    if (this.cacheType === 'redis') {
+      if (!config.redisUrl) throw new Error('redisUrl not set in config');
+      this.cacheEnabled = true;
+      this.client = new Redis(config.redisUrl);
+      this.client.on('error', (err) => {
+        logger.error('Redis error', err);
+        Sentry.captureException(err);
+      });
+    } else if (this.cacheType === 'memory') {
+      this.cacheEnabled = true;
+      this.memoryCache = new LRUCache({
+        // The in-memory cache is really only suited for development, so we'll
+        // hardcode a relatively low limit here.
+        max: 1000,
+      });
+    } else {
+      throw new Error(`Unknown cache type "${this.cacheType}"`);
+    }
+  }
+
+  set(key: string, value: any, maxAgeMS: number) {
+    if (!this.cacheEnabled) return;
+
+    const scopedKey = this.cacheKeyPrefix + key;
+
+    switch (this.cacheType) {
+      case 'memory': {
+        this.memoryCache?.set(scopedKey, JSON.stringify(value), { ttl: maxAgeMS });
```

Comment:
Rather than using optional chaining here and in similar places, let' instead assert that `this.memoryCache` or `this.client` are set when expected:

```ts
import assert from 'node:assert';

assert(this.memoryCache, "Memory cache is enabled but not configured");
```

In practice this shouldn't matter, but this is a good learning experience for defensive programming!

Reply:
I didn't know we could do this! Added assertions and removed optional chaining in [dacc892](https://github.com/PrairieLearn/PrairieLearn/pull/9360/commits/dacc892c32a53e09450e4b8ba921a9242e3bce7f).